### DISCUSSION TITLE
how do you generate the release source file?

### DISCUSSION THREAD

**Discussion by fossfreedom** [Ulauncher/Ulauncher#902] [Packaging]
The source file https://github.com/Ulauncher/Ulauncher/releases/download/5.2.0.r0/ulauncher_5.2.0.r0.tar.gz has the folder structure ./data/preferences/dist/static/js which doesnt exist in the master or dev branches

Creating a debian package it immediately complains about the line length of the js files in the source tar file.  Since this folder structure appears to be generated as part of the release process I was wondering why do you generate this?  Shouldn't this be part of the setup build process?

My understanding (maybe flawed assumption!) is that the source.tar.gz is effectively the snapshot of the branch at the time of the release and shouldnt have built sources.

EDIT:

similar to the above - the release source file has generated ulauncher/ulauncher/utils/desktop/__pycache__/ folders - again, not sure why since any python cache stuff should be built at build time and would vary for the architecture being built on.

**Comment by gornostal** [comment]
Steps to generate a source file:

1. `./ul dev-container`
1. Run this from within the container: `./ul build-deb 5.2.0 --deb`

Then you'll have source files in `/tmp` dir of the container.

**Comment by gornostal** [comment]
`./data/preferences/dist/static/js` doesn't exist in git because `dist` is built by nodejs (webpack to be more accurate) from source files of the VueJS app in `data/preferences/src`

> I was wondering why do you generate this?

This is Preferences UI, which is a VueJS app.

> Shouldn't this be part of the setup build process?

What do you mean by that?

> My understanding (maybe flawed assumption!) is that the source.tar.gz is effectively the snapshot of the branch at the time of the release and shouldnt have built sources.

![image](https://user-images.githubusercontent.com/1202543/58303523-26098e80-7df9-11e9-84b7-52baf4d6140c.png)

Source code (tar.gz) -- is indeed what you mentioned
ulauncher_5.2.0.r0.tar.gz -- is a source code + built preferences UI (i.e. includes `data/preferences/dist`)

> ulauncher/ulauncher/utils/desktop/pycache/ 

I don't see pycache there. What are the steps to reproduce this?

**Comment by fossfreedom** [comment]
ah - ok - my bad - I took the ulauncher_5.2.0.r0.tar.gz and unpacked that - I should have used the "Source code (tar.gz).

apologies for the confusion.

Ok - when building the deb package (I'm not using your method) I cannot use a docker container - the build needs to be from the source tree itself.

First issue on building:

    E: ulauncher source: source-is-missing data/preferences/src/api/fixture.js line length is 2438 characters (>512)

"fixture.js" --> I think the above lintian error is complaining about this bit of the code:

````
            stacktrace: 'Traceback (most recent call last):\n  File "/home/agornostal/projects/ulauncher/ulauncher/api/serve
r/GithubExtension.py", line 101, in _read_json\n    return json.loads(urlopen(url).read().decode(\'utf-8\'))\n  File "/usr/l
ib/python3.6/urllib/request.py", line 223, in urlopen\n    return opener.open(url, data, timeout)\n  File "/usr/lib/python3.
6/urllib/request.py", line 532 ....
````

Is this supposed to be in the source file?  Its hard-coded to your own home project folder

The second question though - can the project be built without docker?  I'm doubtful that this is possible to use docker with an automated debian package.

**Comment by gornostal** [comment]
> Is this supposed to be in the source file? Its hard-coded to your own home project folder

It doesn't affect anything. It's just a function which is used for debugging Preferences UI in a browser. 
I've removed that path.

> when building the deb package (I'm not using your method) I cannot use a docker container - the build needs to be from the source tree itself.

I think now I understand the main problem.
Currently I'm building preferences separately from the debian build. Instead, `debian/control` should have `nodejs` and `yarn` in `Build-Depends`, and then I need to put `yarn install` and `yarn build` somewhere. Probably in `debian/rules`?

**Comment by fossfreedom** [comment]
sure - something like:

```
override_dh_auto_configure:
	yarn install
	yarn build
	dh_auto_configure
```

However typing `yarn install` or `yarn build` in the root of the source tree says:

```
dad:ulauncher-5.2.0$ yarn install
00h00m00s 0/0: : ERROR: [Errno 2] No such file or directory: 'install'
```

**Comment by gornostal** [comment]
@fossfreedom  It works with this:

```
override_dh_auto_configure:
	cd data/preferences && yarn install && yarn build
	dh_auto_configure
```

**Comment by fossfreedom** [comment]
hmm.  Not working for me.

    wget https://github.com/Ulauncher/Ulauncher/archive/5.2.0.r0.tar.gz
    tar -xvf 5.2.0.ro.tar.gz
    cd Ulauncher-5.2.0.r0/data/preferences
    yarn install

same error

Does yarn require some extra parameters that you already setup in your development environment?

**Comment by gornostal** [comment]
You have to use source code tar.gz

**Comment by fossfreedom** [comment]
yep - thats the wget tarball https://github.com/Ulauncher/Ulauncher/archive/5.2.0.r0.tar.gz I used above

**Comment by gornostal** [comment]
hmm weird it works for me. Here's what I do:

1. download https://github.com/Ulauncher/Ulauncher/archive/5.2.0.r0.tar.gz
1. edit `debian/changelog` and replace %VERSION% with 5.2.0
1. edit `debian/rules` and add this in the end  
   ```
   override_dh_auto_configure:
   	cd data/preferences && yarn install && yarn build
   	dh_auto_configure
   ```
1. run `dpkg-buildpackage -tc -us -sa`. In the end it says "error: failed to sign .buildinfo file", but I see the .changes and .dsc files on one level up nevertheless

**Comment by fossfreedom** [comment]
I'm guessing that debian and ubuntu's default yarn version is too old?  v0.32

Is there somehow a way to replace yarn with npm?

**Comment by gornostal** [comment]
Try this instead: `cd data/preferences && npm i yarn && ./node_modules/.bin/yarn && ./node_modules/.bin/yarn build`

I'm currently working on version improvements in this branch https://github.com/Ulauncher/Ulauncher/pull/381/files
I've made some fixes to `debian/control` and `debian/rules` there

**Comment by fossfreedom** [comment]
hmm,

I'm building on Ubuntu 18.04.  This is the results of a local build:

```
dh build --with python3 --buildsystem=pybuild
   dh_update_autotools_config -O--buildsystem=pybuild
   dh_autoreconf -O--buildsystem=pybuild
   debian/rules override_dh_auto_configure
make[1]: Entering directory '/home/dad/Downloads/ulauncher/ulauncher-5.2.0'
cd data/preferences && npm i yarn && ./node_modules/.bin/yarn && ./node_modules/.bin/yarn build
npm WARN prefer global yarn@1.16.0 should be installed with -g
ulauncher-prefs@3.0.0 /home/dad/Downloads/ulauncher/ulauncher-5.2.0/data/preferences
└── yarn@1.16.0  extraneous

yarn install v1.16.0
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@1.2.4: The platform "linux" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning "bootstrap-vue > bootstrap@4.2.1" has unmet peer dependency "jquery@1.9.1 - 3".
warning " > eslint-plugin-prettier@3.0.0" has incorrect peer dependency "eslint@>= 5.0.0".
warning " > eslint-plugin-vue@5.0.0" has incorrect peer dependency "eslint@^5.0.0".
warning " > vue-eslint-parser@4.0.3" has incorrect peer dependency "eslint@^5.0.0".
warning " > url-loader@1.1.2" has incorrect peer dependency "webpack@^3.0.0 || ^4.0.0".
[5/5] Building fresh packages...
Done in 36.37s.
/bin/sh: 1: ./node_modules/.bin/yarn: not found
debian/rules:8: recipe for target 'override_dh_auto_configure' failed
make[1]: *** [override_dh_auto_configure] Error 127
make[1]: Leaving directory '/home/dad/Downloads/ulauncher/ulauncher-5.2.0'
debian/rules:5: recipe for target 'build' failed
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
debuild: fatal error at line 1152:
dpkg-buildpackage -rfakeroot -us -uc -ui failed
dad:ulauncher-5.2.0$ cd data/preferences/node_modules/.
./    ../   .bin/ 
dad:ulauncher-5.2.0$ cd data/preferences/node_modules/.bin/
dad:.bin$ ls
acorn         dateformat                    extract-zip    jsesc         _mocha          phantomjs    sshpk-sign
ansi-html     editorconfig                  handlebars     json5         node-gyp        prettier     sshpk-verify
atob          errno                         he             js-yaml       node-sass       regjsparser  strip-indent
babylon       escodegen                     html-beautify  karma         nopt            rimraf       svgo
browserslist  esgenerate                    html-minifier  loose-envify  not-in-install  sassgraph    uglifyjs
cross-env     eslint                        in-install     miller-rabin  not-in-publish  semver       uuid
css-beautify  eslint-config-prettier-check  in-publish     mime          oc              sha.js       webpack
cssesc        esparse                       istanbul       mkdirp        opencollective  shjs         webpack-bundle-analyzer
csso          esvalidate                    js-beautify    mocha         opener          sshpk-conv   which
dad:.bin$ 
```

I tested the disco build via sbuild (This is how packages are built on Debian and Ubuntu's Launchpad) and the results were

```
dh build --with python3 --buildsystem=pybuild
   dh_update_autotools_config -O--buildsystem=pybuild
   dh_autoreconf -O--buildsystem=pybuild
   debian/rules override_dh_auto_configure
make[1]: Entering directory '/<<PKGBUILDDIR>>'
cd data/preferences && npm i yarn && ./node_modules/.bin/yarn && ./node_modules/.bin/yarn build
npm WARN npm npm does not support Node.js v10.15.2
npm WARN npm You should probably upgrade to a newer version of node as we
npm WARN npm can't make any promises that npm will work with this version.
npm WARN npm Supported releases of Node.js are the latest release of 4, 6, 7, 8, 9.
npm WARN npm You can find the latest version at https://nodejs.org/
Unhandled rejection Error: EACCES: permission denied, mkdir '/sbuild-nonexistent'
npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
make[1]: *** [debian/rules:8: override_dh_auto_configure] Error 1
make[1]: Leaving directory '/<<PKGBUILDDIR>>'
make: *** [debian/rules:5: build] Error 2
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
```

I presume npm i yarn is trying to download stuff from somewhere outside the sbuild sandbox?  i.e. dont think its possible to download off-site software for builds.

**Comment by gornostal** [comment]
> I presume npm i yarn is trying to download stuff from somewhere outside the sbuild sandbox?

Yes, this is what it's design to do. It's like `pip` but for nodejs.

> don't think its possible to download off-site software for builds.

If that's the case then the approach with building Ulauncher from the source code is not going to work. We must have pre-built JS files before running `sbuild`

**Comment by gornostal** [comment]
```
E: ulauncher source: source-is-missing data/preferences/src/api/fixture.js line length is 2438 characters (>512)
```
This issue is fixed in `dev` branch, as I removed that long hard-coded string. 
I'll attach a new tar.gz with pre-built JS files shortly...

**Comment by gornostal** [comment]
[ulauncher.tar.gz](https://github.com/Ulauncher/Ulauncher/files/3219999/ulauncher.tar.gz)

I have only these warnings when building from the source attached above
```
W: ulauncher: priority-extra-is-replaced-by-priority-optional                                                                                                                                                                          
W: ulauncher: duplicate-font-file usr/share/ulauncher/preferences/dist/static/fonts/FontAwesome.otf also in fonts-glewlwyd                                                                                                             
W: ulauncher: duplicate-font-file usr/share/ulauncher/preferences/dist/static/fonts/fontawesome-webfont.ttf also in fonts-glewlwyd                                                                                                     
W: ulauncher: binary-without-manpage usr/bin/ulauncher                                                                                                                                                                                 
W: ulauncher: binary-without-manpage usr/bin/ulauncher-toggle                                                                                                                                                                          
W: ulauncher: desktop-command-not-in-package usr/share/applications/ulauncher.desktop env 
```

UPD: re-uploaded the file

**Comment by gornostal** [comment]
@fossfreedom could you tell what command you run in order to build via sbuild, please?

This one?
```
dh build --with python3 --buildsystem=pybuild \
            dh_update_autotools_config -O--buildsystem=pybuild \
            dh_autoreconf -O--buildsystem=pybuild \
            debian/rules override_dh_auto_configure
```

**Comment by fossfreedom** [comment]
So for sbuild - it is setup as per this wiki https://wiki.ubuntu.com/SimpleSbuild

I have built it using the debian rules 

```
#!/usr/bin/make -f
export DH_VERBOSE = 1

%:
	dh $@ --with python3 --buildsystem=pybuild
```

i.e. `debuild -S -uc -d` to create the source  package
then `sbuild -d disco`

It does sound like that your `make dist` process is needed now since Debian/Ubuntu have a too-old version of `yarn` and hence I'll need to switch to using the dist tarball that you also publish

I will have to use a linitian override to overcome the `E: ulauncher source: source-is-missing data/preferences/dist/static/js/app.c97c9501ce8e7f75c6d7.js line length is 1065 characters (>512)` issue

Can I request though that your `make dist` process to generate the tarball delete all the __pycache__ folders? i.e. generated python cache files shouldn't be in a source tarball - they should be created as part of the .deb/.rpm binary package production.

```
ulauncher/ulauncher/ui/windows/__pycache__/
ulauncher/ulauncher/ui/__pycache__/
ulauncher/ulauncher/__pycache__/
ulauncher/ulauncher/api/server/__pycache__/
ulauncher/ulauncher/api/shared/item/__pycache__/
ulauncher/ulauncher/api/shared/__pycache__/
ulauncher/ulauncher/api/shared/action/__pycache__/
ulauncher/ulauncher/api/__pycache__/
ulauncher/ulauncher/api/client/__pycache__/
ulauncher/ulauncher/search/calc/__pycache__/
ulauncher/ulauncher/search/__pycache__/
ulauncher/ulauncher/search/file_browser/__pycache__/
ulauncher/ulauncher/search/file_browser/alt_menu/__pycache__/
ulauncher/ulauncher/search/apps/__pycache__/
ulauncher/ulauncher/search/shortcuts/__pycache__/
ulauncher/ulauncher/utils/decorator/__pycache__/
ulauncher/ulauncher/utils/db/__pycache__/
ulauncher/ulauncher/utils/__pycache__/
ulauncher/ulauncher/utils/desktop/__pycache__/
```

**Comment by gornostal** [comment]
Thanks for the info.

I'll make sure pycache folders are not included in the tarball
